### PR TITLE
Eternl and Lode E2E on preview

### LINKED ISSUE #1007: Add Eternl to E2E test suite
- merge preprod PR - https://github.com/Plutonomicon/cardano-transaction-lib/pull/944 with https://github.com/Plutonomicon/cardano-transaction-lib/pull/794

### PR DESCRIPTION
**Pull request by klntsky**
Closes #1007 #1030

### Pre-review checklist

- [ ] All code has been formatted using our config (`make format`)
- [ ] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

**Comment by klntsky** [comment]
We want to improve `delayUntilUtxosSynchronised` a bit - wait only for utxos that are being spent.

File: src/BalanceTx/BalanceTx.purs [hunk 1]
```diff
@@ -105,29 +107,36 @@ balanceTx
...
-  utxos <- ExceptT $ utxosAt ownAddr
-    <#> note CouldNotGetUtxos
+balanceTxWithAddress ownAddrs unbalancedTx = runExceptT do
+  changeAddr <- ExceptT $ pure $ note CouldNotGetWalletAddress $ head ownAddrs
+
+  utxos <- ExceptT $ traverse utxosAt ownAddrs <#>
+    ( traverse (note CouldNotGetUtxos) --Maybe -> Either and unwrap UtxoM
+
+        >>> map (foldr Map.union Map.empty) -- merge all utxos into one map
+    )
```
Comment (jy14898) [inline src/BalanceTx/BalanceTx.purs:131]:
```suggestion
  utxos <- ExceptT $ traverse utxosAt ownAddrs <#>
    traverse (note CouldNotGetUtxos) --Maybe -> Either and unwrap UtxoM

        >>> map (foldr Map.union Map.empty) -- merge all utxos into one map
```

File: src/Contract/Address.purs [hunk 2]
```diff
@@ -142,11 +143,16 @@ getWalletCollateral = do
...
+-- https://github.com/Plutonomicon/cardano-transaction-lib/issues/1045
+ownPaymentPubKeyHash = wrapContract
+  (QueryM.ownPaymentPubKeyHashes <#> (_ >>= head))
 
 -- | Gets the wallet `PubKeyHash` via `getWalletAddress`.
 ownPubKeyHash :: forall (r :: Row Type). Contract r (Maybe PubKeyHash)
-ownPubKeyHash = wrapContract QueryM.ownPubKeyHash
+-- TODO: change this to Maybe (Array Address)
+-- https://github.com/Plutonomicon/cardano-transaction-lib/issues/1045
+ownPubKeyHash = wrapContract (QueryM.ownPubKeyHashes <#> (_ >>= head))
```
Comment (jy14898) [inline src/Contract/Address.purs:155]: I think this comment was meant for `getWalletAddress`, although it is applicable here (just not the type)

File: src/Contract/Test/E2E/Helpers.purs [hunk 3]
```diff
@@ -71,14 +75,17 @@ typeInto selector text page = toAffE $ _typeInto selector text page
 -- | Find the popup page of the wallet. This works for both Nami and Gero
 -- | by looking for a page with a button. If there is a button on the main page
 -- | this needs to be modified.
-findWalletPage :: ExtensionId -> Toppokki.Browser -> Aff (Maybe Toppokki.Page)
```
Comment (jy14898) [inline src/Contract/Test/E2E/Helpers.purs:74]: It seems `ExtensionId` is no longer used, but the newtype is still defined - should it be deleted?

**Review by jy14898 — COMMENTED** [review]
Partial review

File: src/Contract/Test/E2E/Helpers.purs (hunk 3 above)
Comment (klntsky) [reply src/Contract/Test/E2E/Helpers.purs:74]: It is used, e.g. in `inWalletPageOptional`

**Review by jy14898 — COMMENTED** [review]
Some small changes but other than that LGTM, changelog needs addressing

File: src/Contract/Test/E2E/Helpers.purs [hunk 4]
```diff
@@ -201,28 +222,87 @@ derive instance Newtype ExtensionId _
...
+inWalletPageOptional
+  :: forall (a :: Type)
+   . ExtensionId
+  -> Pattern
+  -> RunningExample
+  -> Number
+  -> (Toppokki.Page -> Aff a)
+  -> Aff (Maybe a)
+inWalletPageOptional extId pattern { browser, jQuery } timeout cont = do
+  try acquirePage <#> fromRight Nothing
```
[the remark this answers is not in the document]
Comment (klntsky) [reply src/Contract/Test/E2E/Helpers.purs:248]: `fromRight Nothing` /= `hush`

File: src/QueryM/Utxos.purs [hunk 5]
```diff
@@ -155,9 +184,11 @@ getWalletUtxos = do
     Gero wallet -> liftAff $ wallet.getUtxos wallet.connection <#> map toUtxoMap
     Flint wallet -> liftAff $ wallet.getUtxos wallet.connection <#> map
       toUtxoMap
+    Eternl wallet -> liftAff $ wallet.getUtxos wallet.connection <#> map
+      toUtxoMap
     Lode wallet -> liftAff $ wallet.getUtxos wallet.connection <#> map toUtxoMap
     KeyWallet _ -> do
-      mbAddress <- getWalletAddress
+      mbAddress <- (getWalletAddresses <#> (_ >>= head))
```
Comment (ngua) [inline src/QueryM/Utxos.purs:191]: Redundant parens here

**Review by ngua — APPROVED** [review]
Just a few small comments on formatting, otherwise looks good

**Comment by ngua** [comment]
I'll fix the merge conflicts now

**Pull request merged**